### PR TITLE
Issue 3191 expose isbullet

### PR DESCRIPTION
**Pull request by otsakir**
Introduces 'bullet' property to collisionobject. See #3191.

### Implementation

Property added only for 2D physics - Box2D. In 3D physics if you try to get the 'bullet' property the following  warning/error is displayed
```
WARNING:GAMESYS: 'bullet' property not supported in 3d physics mode
ERROR:SCRIPT: /main/bullet.script:38: '/go#collisionobject' does not have any property called 'bullet'
stack traceback:
  [C]:-1: in function get
  /main/bullet.script:38: in function </main/bullet.script:26>
```

### Testing

The feature is tested by getting and setting the 'bullet' property and asserting it is valid. No  test code is written to test actual behavior of a bullet sprite and whether this is accomplished. This is assumed to be part part of box2d testing. Also, you'll notice that i've renamed `Sleeping2DCollisionObjectTest` to  `CollisionObject2DTest`. The idea is to use this test class as the base for generic high-level collision object testing in gamesys module. Also, `properties.script` is supposed to be the test script where elementary testing is done for physics properties. 

I've also built the editor with bullet property support, created a new project that plays with the property and assured it is shown and initialized properly while getters setters also work. Everything seems to work ok.

### Editor

After the core implementation (writing code that makes the calls to box2d and provide the lua api) and test, some further changes were needed to make the property work seamlessly, allow initialization from a .collisionobject file, get displayed in the editor. I used 'Locked Rotation' property as a guide.

### CODE REVIEW

**Comment by britzl** [comment]
Fantastic! Is the feature ready for review?

**Comment by otsakir** [comment]
@JCash , please tell me if you need any more help with this PR. I noticed there is a test failing.

File: engine/gamesys/src/gamesys/components/comp_collision_object.cpp [hunk 1]
```diff
@@ -1160,6 +1164,15 @@ namespace dmGameSystem
                 out_value.m_Variant = dmGameObject::PropertyVar(dmPhysics::GetMass2D(component->m_Object2D));
             }
             return dmGameObject::PROPERTY_RESULT_OK;
+        } else if (params.m_PropertyId == PROP_BULLET) {
+            if (physics_context->m_3D) {
+                // TODO - display not-implemented warning
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_collision_object.cpp:1169]: You've done the TODO, so you can remove the comment :)

File: engine/gamesys/src/gamesys/test/test_gamesys.cpp [hunk 2]
```diff
@@ -1438,7 +1438,7 @@ TEST_F(GamepadConnectedTest, TestGamepadConnectedInputEvent)
     dmGameSystem::FinalizeScriptLibs(scriptlibcontext);
 }
 
-TEST_F(Sleeping2DCollisionObjectTest, WakingCollisionObjectTest)
+TEST_F(CollisionObject2DTest, WakingCollisionObjectTest)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/test/test_gamesys.cpp:1441]: As shown in the previous physics PR, we also need to add a "compile time" test in `test_physics.cpp`.
This is needed to make sure we catch missing features early on.
In this case, there is no change to `physics_2d_null.cpp`, which is required.

**Review by JCash — CHANGES_REQUESTED** [review]
Looks good.
Missing test + implementation for null implementation.

File: engine/gamesys/src/gamesys/test/test_gamesys.cpp (hunk 2 above)
Comment (otsakir) [reply engine/gamesys/src/gamesys/test/test_gamesys.cpp:1441]: @JCash , please elaborate a little here. wdym by 'compile time' test ? A test that will involve calling  dmPhysics::SetBullet2d/IsBullet2D() that will fail if the feature is not present ?

I went through physics_2d_null.cpp. I noticed that it contains dummy implementations of functionality typically present in physics_2d.cpp. 

Could you briefly explain how this works ?

Comment (otsakir) [reply engine/gamesys/src/gamesys/test/test_gamesys.cpp:1441]: Nevermind. We solved this.

**Comment by otsakir** [comment]
@JCash , i've added "implementation" in physics_2d_null.cpp. 
* Is there some sort of test i need to add ?  I think you mentioned something about test_physics_2d.cpp. Maybe something that just creates a collision object and accesses Bullet* operations ?
* In the test build directory i noticed that there are both `test_physics_2d` and `test_physics_2d_null`  binaries produced. Why have a "test_physics_2d_null"? What is the behavior i shold  expect  when running these ?

**Comment by JCash** [comment]
@otsakir As I mentioned in the Discord chat, our users use the `physics_null` and `physics_2d_null` libraries too.
So if we don't implement the functions there too, they will get `Undefined symbols` errors when building the game.
An example of a fix for it is [here](https://github.com/defold/defold/pull/5878).

Since it's easy to forget, we also added a unit test specifically for this. they are called `test_physics_null` and `test_physics_2d_null`.

These tests link against the `physics_null` and `physics_2d_null` libraries respectively. So, if you're using a new function in a test, but forget to add the null implementation, it is automatically caught and you'll get a linker error (Undefined symbol).

Try it!
Comment out your null implementation, add a test using the physics function, then build the physics tests, and you should get a linker error.

File: engine/physics/src/physics/test/test_physics_2d.cpp [hunk 3]
```diff
@@ -175,6 +177,28 @@ TYPED_TEST(PhysicsTest, MultipleGroups)
...
+    dmPhysics::CollisionObjectData data;
+    data.m_Type = dmPhysics::COLLISION_OBJECT_TYPE_DYNAMIC;
+    data.m_Mass = 1.0f;
+    data.m_UserData = &vo_b;
+    data.m_Group = 1 << 3;
+    data.m_Mask = 1 << 2;
+    typename TypeParam::CollisionShapeType shape = (*TestFixture::m_Test.m_NewBoxShapeFunc)(TestFixture::m_Context, Vector3(0.5f, 0.5f, 0.0f));
+    typename TypeParam::CollisionObjectType dynamic_co = (*TestFixture::m_Test.m_NewCollisionObjectFunc)(TestFixture::m_World, data, &shape, 1u);
+
+    (*TestFixture::m_Test.m_IsBulletFunc)(dynamic_co);
```
Comment (JCash) [inline engine/physics/src/physics/test/test_physics_2d.cpp:194]: Is there a reason you don't check the return values of the function?

File: engine/physics/src/physics/test/test_physics_2d.cpp [hunk 4]
```diff
@@ -175,6 +177,28 @@ TYPED_TEST(PhysicsTest, MultipleGroups)
...
+    (*TestFixture::m_Test.m_SetBulletFunc)(dynamic_co, false);
```
Comment (JCash) [inline engine/physics/src/physics/test/test_physics_2d.cpp:195]: SInce the default is `false`, isn't it better to test to set it to `true`?

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/physics/src/physics/test/test_physics_2d.cpp (hunk 4 above)
Comment (otsakir) [reply engine/physics/src/physics/test/test_physics_2d.cpp:195]: I thought the test was mostly about calling the functions and assuring they don't produce a link error. Let me fix...

File: engine/physics/src/physics/test/test_physics_2d.cpp (hunk 3 above)
Comment (otsakir) [reply engine/physics/src/physics/test/test_physics_2d.cpp:194]: I thought the test was mostly about calling the functions and assuring they don't produce a link error. Let me fix...

**Comment by otsakir** [comment]
@JCash , ready for review!
